### PR TITLE
Setup initial version of password-store.

### PR DESCRIPTION
**Pull request by rbartlensky**
Currently it supports:
 * storing passwords on local keyring
 * getting passwords from local keyring and nitrokey

### CODE REVIEW

File: src/config.rs [hunk 1]
```diff
@@ -0,0 +1,35 @@
+use crate::error::PasswordStoreError;
+use std::fs::{create_dir, File};
+use std::io::{Read, Write};
+use std::path::Path;
+use toml::{map::Map, Value};
+
+const DEFAULTS_PATH: &'static str = "/home/robert/.config/password-store-rust";
+const TOML_NAME: &'static str = "defaults.toml";
```
Comment (gabi-250) [inline src/config.rs]: I suggest renaming this to `DEFAULTS_TOML`.

File: src/config.rs [hunk 2]
```diff
@@ -0,0 +1,35 @@
+use crate::error::PasswordStoreError;
+use std::fs::{create_dir, File};
+use std::io::{Read, Write};
+use std::path::Path;
+use toml::{map::Map, Value};
+
+const DEFAULTS_PATH: &'static str = "/home/robert/.config/password-store-rust";
```
Comment (gabi-250) [inline src/config.rs]: Don't hard-code the home directory.

File: src/config.rs [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+    if !defaults_path.exists() {
+        create_dir(DEFAULTS_PATH)?;
+    }
+    let defaults_file = defaults_path.join(TOML_NAME);
+    if !defaults_file.exists() {
+        File::create(defaults_file.clone())?;
+    }
+    let mut defaults = File::open(defaults_file)?;
+    let mut toml_payload = String::new();
+    defaults.read_to_string(&mut toml_payload)?;
```
Comment (gabi-250) [inline src/config.rs]: Use `std::fs::read_to_string` instead.

File: src/config.rs [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+        .ok()
+        .and_then(|r| match r {
+            Value::Table(table) => Some(table),
+            _ => None,
+        })
+        .unwrap_or(Map::new()))
+}
+
+pub fn write_toml(toml: Map<String, Value>) -> Result<(), PasswordStoreError> {
+    let mut config_file = File::create(Path::new(DEFAULTS_PATH).join(TOML_NAME))?;
```
Comment (gabi-250) [inline src/config.rs]: Use `std::fs::write` instead.

File: src/get.rs [hunk 5]
```diff
@@ -0,0 +1,72 @@
+use crate::config::get_config_toml;
+use crate::error::PasswordStoreError;
+use clap::Values;
+use keyring::Keyring;
+use nitrokey::GetPasswordSafe;
+use nitrokey::{connect, SLOT_COUNT};
```
Comment (gabi-250) [inline src/get.rs]: You could group these under `nitrokey::`.

File: src/config.rs [hunk 6]
```diff
@@ -0,0 +1,35 @@
+use crate::error::PasswordStoreError;
+use std::fs::{create_dir, File};
+use std::io::{Read, Write};
+use std::path::Path;
+use toml::{map::Map, Value};
+
+const DEFAULTS_PATH: &'static str = "/home/robert/.config/password-store-rust";
+const TOML_NAME: &'static str = "defaults.toml";
+
+pub fn get_config_toml() -> Result<Map<String, Value>, PasswordStoreError> {
```
Comment (gabi-250) [inline src/config.rs]: Maybe rename this to `parse_config`.

File: src/get.rs [hunk 7]
```diff
@@ -0,0 +1,72 @@
...
+use rpassword::prompt_password_stdout;
+use std::io::Write;
+use std::process::{Command, Stdio};
+use std::thread::sleep;
+use std::time::Duration;
+
+fn get_default_username(service: &str) -> Result<String, PasswordStoreError> {
```
Comment (gabi-250) [inline src/get.rs]: `get_default_username` -> `default_username`

File: src/get.rs [hunk 8]
```diff
@@ -0,0 +1,72 @@
...
+    let toml = get_config_toml()?;
+    match toml.get(service) {
+        Some(username) => Ok(username.as_str().unwrap().into()),
+        None => Err(PasswordStoreError::NoDefaultUser(
+            format!("No default set for {}", service).into(),
+        )),
+    }
+}
+
+fn xclip(payload: &str) -> Result<(), PasswordStoreError> {
```
Comment (gabi-250) [inline src/get.rs]: It's probably a better idea to use [clipboard](https://crates.io/crates/clipboard).

File: src/get.rs [hunk 9]
```diff
@@ -0,0 +1,72 @@
...
+        .stdin(Stdio::piped())
+        .spawn()?;
+    xclip.stdin.unwrap().write_all(payload.as_bytes())?;
+    Ok(())
+}
+
+pub fn get_password_from_nitrokey<'a>(service: &'a str, username: &'a str) -> Option<String> {
+    let device = match connect() {
+        Ok(device) => device,
+        _ => {
```
Comment (gabi-250) [inline src/get.rs]: Should this function return a `Result` instead?

File: src/get.rs [hunk 10]
```diff
@@ -0,0 +1,72 @@
...
+            return None;
+        }
+    };
+    let pin = prompt_password_stdout("Nitro-key user pin: ").unwrap();
+    let password_safe = device.get_password_safe(&pin).unwrap();
```
Comment (gabi-250) [inline src/get.rs]: We probably need an error type for nitrokey-related errors (e.g. wrong pin).

File: src/get.rs [hunk 11]
```diff
@@ -0,0 +1,72 @@
...
+pub fn handle_get(mut values: Values) -> Result<(), PasswordStoreError> {
+    let service = values.next().unwrap();
+    let username = match values.next() {
+        Some(username) => username.into(),
+        None => get_default_username(service)?,
+    };
+    // search nitro-key for password
+    let password = match get_password_from_nitrokey(service, &username) {
+        Some(pw) => pw,
+        None => {
```
Comment (gabi-250) [inline src/get.rs]: It might be worth printing a message to indicate where the password was retrieved from.

File: src/main.rs [hunk 12]
```diff
@@ -0,0 +1,39 @@
+use clap::*;
+mod error;
+mod get;
+use get::handle_get;
+mod set;
+use set::handle_set;
+mod config;
+
+fn main() {
+    let matches = App::new("Password store")
```
Comment (gabi-250) [inline src/main.rs]: `Password store` sounds a little generic :-).

File: src/main.rs [hunk 13]
```diff
@@ -0,0 +1,39 @@
...
+        .version("1.0")
+        .author("Robert B. <bartlensky.robert@gmail.com>")
+        .about("Stores passwords on the local keyring or on a nitro key.")
+        .arg(
+            Arg::with_name("get")
```
Comment (gabi-250) [inline src/main.rs:16]: I think this should be `service`, not `get`.

File: src/main.rs [hunk 14]
```diff
@@ -0,0 +1,39 @@
...
+                .short("g")
+                .long("get")
+                .help("Copy the password for <service> to the kill-ring.")
+                .takes_value(true)
+                .min_values(1)
+                .max_values(2),
+        )
+        .arg(
+            Arg::with_name("set")
```
Comment (gabi-250) [inline src/main.rs:25]: `set` -> `service`

File: src/set.rs [hunk 15]
```diff
@@ -0,0 +1,33 @@
+use crate::config::{get_config_toml, write_toml};
+use crate::error::PasswordStoreError;
+use clap::Values;
+use keyring::Keyring;
+use std::process::Command;
+use std::str::from_utf8;
+use toml::Value;
+
+fn get_random_password() -> Result<String, PasswordStoreError> {
+    let pass = Command::new("pwgen")
```
Comment (gabi-250) [inline src/set.rs]: Consider using an external crate to generate passwords. [passwords](https://crates.io/crates/passwords) is fairly similar to `pwgen`.

File: src/set.rs [hunk 16]
```diff
@@ -0,0 +1,33 @@
...
+    let service = values.next().unwrap();
+    let username = values.next().unwrap();
+    let password = match values.next() {
+        Some(password) => password.into(),
+        None => get_random_password()?,
+    };
+    let keyring = Keyring::new(service, username);
+    keyring.set_password(&password)?;
+    let mut toml = get_config_toml()?;
+    if let None = toml.get(service) {
```
Comment (gabi-250) [inline src/set.rs]: I prefer `toml.get(service).is_none()`, but it's entirely up to you.

**Review by gabi-250 — CHANGES_REQUESTED** [review]
Looks like a good start! 

A few other comments:
* perhaps list the dependencies in the README (e.g. libhidapi-dev)
* `nitro-key` -> `nitrokey`

**Comment by rbartlensky** [comment]
@gabi-250 Thank you for your comments. I think the last commit should address everything you mentioned!

**Review by gabi-250 — APPROVED** [review]
Looks good to me. Do you want to squash?

**Comment by rbartlensky** [comment]
Squashed!

**Pull request merged**